### PR TITLE
Add SimMIM

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds an `AutoModelForMaskedImageModeling` API to the library with a corresponding `run_mim.py` script. 

It allows to pre-train the following Transformer-based vision models:

* ViT
* DeiT
* BEiT
* Swin Transformer

based on Microsoft's cool (and simple) [SimMIM](https://arxiv.org/abs/2111.09886) paper (note that this won't work for convolution based models such as ConvNeXT).

This requires a tiny tweak in each modeling file, unfortunately. However, it's quite minimal, namely I added the ability to add a `mask_token` trainable parameter. This is a bit different compared to NLP models, which have the mask token inside their embedding layer (`nn.Embedding`).

The `run_mim.py` script is directly inspired by `run_mlm.py`, but adapted to work for vision.

### CODE REVIEW

File: examples/pytorch/image-pretraining/README.md [hunk 1]
```diff
@@ -16,9 +16,125 @@ limitations under the License.
...
-This directory contains a script, `run_mae.py`, that can be used to pre-train a Vision Transformer as a masked autoencoder (MAE), as proposed in [Masked Autoencoders Are Scalable Vision Learners](https://arxiv.org/abs/2111.06377). The script can be used to train a `ViTMAEForPreTraining` model in the Transformers library, using PyTorch. After self-supervised pre-training, one can load the weights of the encoder directly into a `ViTForImageClassification`. The MAE method allows for learning high-capacity models that generalize well: e.g., a vanilla ViT-Huge model achieves the best accuracy (87.8%) among methods that use only ImageNet-1K data.
+
+# SimMIM
+
+The `run_mim.py` script can be used to pre-train any Transformer-based vision model in the library (concretly, any model supported by the `AutoModelForMaskedImageModeling` API) for masked image modeling (as proposed in [SimMIM: A Simple Framework for Masked Image Modeling](https://arxiv.org/abs/2111.09886)) using PyTorch.
+
+The goal for the model is to predict raw pixel values for the masked patches, using just a linear layer as prediction head. The model is trained using a simple L1 loss.
+
+## Using datasets from 🤗 `datasets`
```
Comment (sgugger) [inline examples/pytorch/image-pretraining/README.md:32]:
```suggestion
## Using datasets from 🤗 Datasets
```

File: examples/pytorch/image-pretraining/README.md [hunk 2]
```diff
@@ -16,9 +16,125 @@ limitations under the License.
...
 
+This directory contains Python scripts that allow you to pre-train Transformers for computer vision tasks on your own data, after which you can easily load the weights into a `ViTForImageClassification` for instance. It currently includes scripts for:
+- [SimMIM](#simmim) (by Microsoft Research)
+- [MAE](#mae) (by Facebook AI).
+
 NOTE: If you encounter problems/have suggestions for improvement, open an issue on Github and tag @NielsRogge.
 
-This directory contains a script, `run_mae.py`, that can be used to pre-train a Vision Transformer as a masked autoencoder (MAE), as proposed in [Masked Autoencoders Are Scalable Vision Learners](https://arxiv.org/abs/2111.06377). The script can be used to train a `ViTMAEForPreTraining` model in the Transformers library, using PyTorch. After self-supervised pre-training, one can load the weights of the encoder directly into a `ViTForImageClassification`. The MAE method allows for learning high-capacity models that generalize well: e.g., a vanilla ViT-Huge model achieves the best accuracy (87.8%) among methods that use only ImageNet-1K data.
+
+# SimMIM
```
Comment (sgugger) [inline examples/pytorch/image-pretraining/README.md:26]: Expand what the title means, as it's very likely no one knows what SimMIM is.

File: examples/pytorch/image-pretraining/README.md [hunk 3]
```diff
@@ -16,9 +16,125 @@ limitations under the License.
...
+DEPTHS = [2, 2, 18, 2]
+NUM_HEADS = [4, 8, 16, 32]
+WINDOW_SIZE = 6
+
+config = SwinConfig(image_size=IMAGE_SIZE,
+                    patch_size=PATCH_SIZE,
+                    embed_dim=EMBED_DIM,
+                    depths=DEPTHS,
+                    num_heads=NUM_HEADS,
+                    window_size=WINDOW_SIZE)
```
Comment (sgugger) [inline examples/pytorch/image-pretraining/README.md:76]: Should be properly formatted.

File: examples/pytorch/image-pretraining/run_mim.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+    """
+
+    def __init__(self, input_size=192, mask_patch_size=32, model_patch_size=4, mask_ratio=0.6):
+        self.input_size = input_size
+        self.mask_patch_size = mask_patch_size
+        self.model_patch_size = model_patch_size
+        self.mask_ratio = mask_ratio
+
+        assert self.input_size % self.mask_patch_size == 0
+        assert self.mask_patch_size % self.model_patch_size == 0
```
Comment (sgugger) [inline examples/pytorch/image-pretraining/run_mim.py:188]: We shoudn't have empty asserts.

File: src/transformers/models/beit/configuration_beit.py [hunk 5]
```diff
@@ -48,22 +48,20 @@ class BeitConfig(PretrainedConfig):
...
+        image_size (`int`, *optional*, defaults to 224):
             The size (resolution) of each image.
-        patch_size (`int`, *optional*, defaults to `16`):
+        patch_size (`int`, *optional*, defaults to 16):
             The size (resolution) of each patch.
-        num_channels (`int`, *optional*, defaults to `3`):
+        num_channels (`int`, *optional*, defaults to 3):
             The number of input channels.
-        use_mask_token (`bool`, *optional*, defaults to `False`):
-            Whether to use a mask token for masked image modeling.
```
Comment (sgugger) [inline src/transformers/models/beit/configuration_beit.py:66]: Removing this argument looks like a breaking change.

File: src/transformers/models/beit/modeling_beit.py [hunk 6]
```diff
@@ -142,14 +142,11 @@ class BeitEmbeddings(nn.Module):
...
-    def __init__(self, config):
+    def __init__(self, config, use_mask_token=False):
         super().__init__()
 
         self.cls_token = nn.Parameter(torch.zeros(1, 1, config.hidden_size))
-        if config.use_mask_token:
-            self.mask_token = nn.Parameter(torch.zeros(1, 1, config.hidden_size))
-        else:
-            self.mask_token = None
+        self.mask_token = nn.Parameter(torch.zeros(1, 1, config.hidden_size)) if use_mask_token else None
```
Comment (sgugger) [inline src/transformers/models/beit/modeling_beit.py:149]: This is breaking since if there is a model on the Hub with a checkpoint that has `use_mask_token=True` in its config, but is then instantiated in a class that doesn't send `use_mask_token=True` to this function, we will have a mistmatch in parameters and the load will crash.

We need to have something that is more backward compatible.

File: src/transformers/models/beit/modeling_beit.py [hunk 7]
```diff
@@ -718,12 +715,18 @@ class BeitForMaskedImageModeling(BeitPreTrainedModel):
...
+        if config.decoder_type == "beit":
+            self.layernorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+            self.lm_head = nn.Linear(config.hidden_size, config.vocab_size)
+        elif config.decoder_type == "simmim":
+            self.decoder = nn.Sequential(
+                nn.Conv2d(in_channels=config.hidden_size, out_channels=config.encoder_stride**2 * 3, kernel_size=1),
+                nn.PixelShuffle(config.encoder_stride),
+            )
+        else:
+            raise ValueError(f"Unknown decoder type {config.decoder_type}. Only 'beit' and 'simmim' are supported.")
```
Comment (sgugger) [inline src/transformers/models/beit/modeling_beit.py:729]: This should probably be raised inside the `config`, so the user gets the error message as early as possible.

File: src/transformers/models/beit/modeling_beit.py [hunk 8]
```diff
@@ -780,20 +778,42 @@ def forward(
...
+
+            if labels is not None:
+                loss_fct = CrossEntropyLoss()  # -100 index = padding token
+                masked_im_loss = loss_fct(prediction_scores[bool_masked_pos], labels)
+        elif self.config.decoder_type == "simmim":
+            # Reshape to (batch_size, num_channels, height, width)
+            sequence_output = sequence_output[:, 1:]
+            B, L, C = sequence_output.shape
+            H = W = int(L**0.5)
+            sequence_output = sequence_output.permute(0, 2, 1).reshape(B, C, H, W)
```
Comment (sgugger) [inline src/transformers/models/beit/modeling_beit.py:795]: B, L, C, H, W are not classes, so should not be capitalized, and should not be one-letter variable names.

File: src/transformers/models/deit/modeling_deit.py [hunk 9]
```diff
@@ -545,6 +549,105 @@ def forward(self, hidden_states):
...
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+
+        # Reshape to (batch_size, num_channels, height, width)
+        sequence_output = sequence_output[:, 1:-1]
+        B, L, C = sequence_output.shape
+        H = W = int(L**0.5)
+        sequence_output = sequence_output.permute(0, 2, 1).reshape(B, C, H, W)
```
Comment (sgugger) [inline src/transformers/models/deit/modeling_deit.py:621]: Same as above.

File: src/transformers/models/swin/inference.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
+import numpy as np
```
Comment (sgugger) [inline src/transformers/models/swin/inference.py:1]: This should be removed.

File: src/transformers/models/swin/modeling_swin.py [hunk 11]
```diff
@@ -757,6 +771,106 @@ def forward(
...
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+
+        # Reshape to (batch_size, num_channels, height, width)
+        sequence_output = sequence_output.transpose(1, 2)
+        B, C, L = sequence_output.shape
+        H = W = int(L**0.5)
+        sequence_output = sequence_output.reshape(B, C, H, W)
```
Comment (sgugger) [inline src/transformers/models/swin/modeling_swin.py:844]: Same as above.

File: src/transformers/models/vit/inference.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import torch
```
Comment (sgugger) [inline src/transformers/models/vit/inference.py:1]: This file should be removed.

File: src/transformers/models/vit/modeling_vit.py [hunk 13]
```diff
@@ -117,10 +123,17 @@ def interpolate_pos_encoding(self, embeddings, height, width):
...
+    def forward(self, pixel_values, bool_masked_pos=None, interpolate_pos_encoding=False):
         batch_size, num_channels, height, width = pixel_values.shape
         embeddings = self.patch_embeddings(pixel_values, interpolate_pos_encoding=interpolate_pos_encoding)
 
+        batch_size, seq_len, _ = embeddings.size()
+        if bool_masked_pos is not None:
+            mask_tokens = self.mask_token.expand(batch_size, seq_len, -1)
+            # replace the masked visual tokens by mask_tokens
+            w = bool_masked_pos.unsqueeze(-1).type_as(mask_tokens)
+            embeddings = embeddings * (1.0 - w) + mask_tokens * w
```
Comment (sgugger) [inline src/transformers/models/vit/modeling_vit.py:135]: Let's use something more informative than `w` as variable name, please.

File: src/transformers/models/vit/modeling_vit.py [hunk 14]
```diff
@@ -422,10 +435,6 @@ def _init_weights(self, module):
             module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
             if module.bias is not None:
                 module.bias.data.zero_()
-        elif isinstance(module, nn.Embedding):
-            module.weight.data.normal_(mean=0.0, std=self.config.initializer_range)
-            if module.padding_idx is not None:
-                module.weight.data[module.padding_idx].zero_()
```
Comment (sgugger) [inline src/transformers/models/vit/modeling_vit.py:428]: Why is this removed?

File: src/transformers/models/vit/modeling_vit.py [hunk 15]
```diff
@@ -573,6 +585,107 @@ def forward(self, hidden_states):
...
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+
+        # Reshape to (batch_size, num_channels, height, width)
+        sequence_output = sequence_output[:, 1:]
+        B, L, C = sequence_output.shape
+        H = W = int(L**0.5)
+        sequence_output = sequence_output.permute(0, 2, 1).reshape(B, C, H, W)
```
Comment (sgugger) [inline src/transformers/models/vit/modeling_vit.py:659]: Same comment as before on naming.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this new example!

This PR tries to do too much at the same time however:
- there is breaking change in BeIT, so this model should be excluded from this PR and added separately once this one is merged, so we can take more action to mitigate the breaking change and be more backward compatible (in particular, the default value for the `use_mask_token` in the embedding should be the one given by the config)
- there is another breaking change in the initialization of ViT which should also go in a separate PR

Pay attention to the file you include in your PR, there are three here that are not necessary:
- a `.DS_Store` file
- models/swin/inference.py
- models/vit/inference.py
You should consider using a GUI for git (like GitKraken) if those were added by mistake.

File: src/transformers/models/vit/modeling_vit.py (hunk 14 above)
Comment (NielsRogge) [reply src/transformers/models/vit/modeling_vit.py:428]: Closing this since no `nn.Embedding` layer is used in modeling_vit.py

File: examples/pytorch/image-pretraining/README.md (hunk 3 above)
Comment (NielsRogge) [reply examples/pytorch/image-pretraining/README.md]: What is a proper formatting here?

**Comment by NielsRogge** [comment]
Updated the PR to not include BEiT.

**Review by sgugger — APPROVED** [review]
Thanks for adapting!

File: .gitignore [hunk 16]
```diff
@@ -160,4 +160,7 @@ tags
 .pre-commit*
 
 # .lock
-*.lock
\ No newline at end of file
+*.lock
+
+# DS_Store (MacOS)
+.DS_Store
```
Comment (LysandreJik) [inline .gitignore:166]: Ok!

File: examples/pytorch/image-pretraining/README.md [hunk 17]
```diff
@@ -16,9 +16,132 @@ limitations under the License.
...
 
+This directory contains Python scripts that allow you to pre-train Transformer-based vision models on your own data, after which you can easily load the weights into a `ViTForImageClassification` for instance. It currently includes scripts for:
+- [SimMIM](#simmim) (by Microsoft Research)
+- [MAE](#mae) (by Facebook AI).
+
 NOTE: If you encounter problems/have suggestions for improvement, open an issue on Github and tag @NielsRogge.
 
-This directory contains a script, `run_mae.py`, that can be used to pre-train a Vision Transformer as a masked autoencoder (MAE), as proposed in [Masked Autoencoders Are Scalable Vision Learners](https://arxiv.org/abs/2111.06377). The script can be used to train a `ViTMAEForPreTraining` model in the Transformers library, using PyTorch. After self-supervised pre-training, one can load the weights of the encoder directly into a `ViTForImageClassification`. The MAE method allows for learning high-capacity models that generalize well: e.g., a vanilla ViT-Huge model achieves the best accuracy (87.8%) among methods that use only ImageNet-1K data.
+
+# SimMIM: A Simple Framework for Masked Image Modeling
```
Comment (LysandreJik) [inline examples/pytorch/image-pretraining/README.md:26]: Usually, there should be only one title (single `#`) per document.
```suggestion
## SimMIM: A Simple Framework for Masked Image Modeling
```

File: examples/pytorch/image-pretraining/README.md [hunk 18]
```diff
@@ -16,9 +16,132 @@ limitations under the License.
...
+    --dataset_name nateraw/image-folder \
+    --train_dir <path-to-train-root> \
+    --output_dir ./outputs/ \
+    --remove_unused_columns False \
+    --label_names bool_masked_pos \
+    --do_train \
+    --do_eval
+```
+
+# MAE
```
Comment (LysandreJik) [inline examples/pytorch/image-pretraining/README.md:142]: Usually, there should be only one title (single `#`) per document.

File: examples/pytorch/image-pretraining/README.md [hunk 19]
```diff
@@ -100,7 +223,7 @@ python run_mae.py \
   - To provide your own validation split in its own directory, you can pass the `--validation_dir <path-to-val-root>` flag.
 
 
-## Sharing your model on 🤗 Hub
+# Sharing your model on 🤗 Hub
```
Comment (LysandreJik) [inline examples/pytorch/image-pretraining/README.md:226]: Usually, there should be only one title (single `#`) per document.

File: src/transformers/models/auto/modeling_auto.py [hunk 20]
```diff
@@ -821,6 +833,13 @@ class AutoModelForAudioXVector(_BaseAutoModelClass):
 AutoModelForAudioXVector = auto_class_update(AutoModelForAudioXVector, head_doc="audio retrieval via x-vector")
 
 
+class AutoModelForMaskedImageModeling(_BaseAutoModelClass):
+    _model_mapping = MODEL_FOR_MASKED_IMAGE_MODELING_MAPPING
+
+
+AutoModelForMaskedImageModeling = auto_class_update(AutoModelForMaskedImageModeling, head_doc="masked image modeling")
```
Comment (LysandreJik) [inline src/transformers/models/auto/modeling_auto.py:840]: Cool :)

**Review by LysandreJik — APPROVED** [review]
Cool, thank you for working on that @NielsRogge, super exciting! Feel free to tweet about it when it's merged as the example will land on `master` immediately :)
